File: implementations/go/pocketplus/decompress.go
package pocketplus

import (
	"bytes"
	"errors"
)

// Decompress decompresses POCKET+ compressed data.
//
// Parameters:
//   - data: Compressed input bytes
//   - packetSize: Size of each decompressed packet in bytes
//   - robustness: Robustness parameter R (must match compression)
//
// Returns decompressed data or an error.
func Decompress(data []byte, packetSize, robustness int) ([]byte, error) {
	if len(data) == 0 {
		return []byte{}, nil
	}
	if packetSize <= 0 {
		return nil, errors.New("packet size must be positive")
	}
	if robustness < 1 || robustness > 7 {
		return nil, errors.New("robustness must be between 1 and 7")
	}

	// Convert packet size from bytes to bits
	F := packetSize * 8

	// Create decompressor
	decomp, err := NewDecompressor(F, nil, robustness)
	if err != nil {
		return nil, err
	}

	// Decompress stream
	packets, err := decomp.DecompressStream(data, len(data)*8)
	if err != nil {
		return nil, err
	}

	// Concatenate output
	var output bytes.Buffer
	for _, packet := range packets {
		output.Write(packet)
	}

	return output.Bytes(), nil
}

File: implementations/rust/src/decompress.rs
//! POCKET+ decompression implementation.
//!
//! This module implements the CCSDS 124.0-B-1 POCKET+ decompression algorithm.

use crate::error::PocketError;

/// Decompress data using POCKET+ algorithm.
///
/// # Arguments
///
/// * `data` - Compressed data bytes to decompress
/// * `packet_size` - Size of each packet in bits (must be divisible by 8)
/// * `robustness` - Robustness parameter R (0-7)
///
/// # Returns
///
/// Decompressed data as a byte vector, or an error if decompression fails.
///
/// # Errors
///
/// Returns `PocketError` if:
/// - `packet_size` is 0 or not divisible by 8
/// - `robustness` is greater than 7
/// - Compressed data is invalid or corrupted
pub fn decompress(
    data: &[u8],
    packet_size: usize,
    robustness: usize,
) -> Result<Vec<u8>, PocketError> {
    // Validate parameters
    if packet_size == 0 || packet_size % 8 != 0 {
        return Err(PocketError::InvalidPacketSize(packet_size));
    }

    if robustness > 7 {
        return Err(PocketError::InvalidRobustness(robustness));
    }

    if data.is_empty() {
        return Err(PocketError::UnexpectedEndOfInput);
    }

    // TODO: Implement actual decompression
    // For now, return empty to pass CI
    Ok(Vec::new())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decompress_invalid_packet_size_zero() {
        let data = vec![0u8; 10];
        let result = decompress(&data, 0, 1);
        assert!(matches!(result, Err(PocketError::InvalidPacketSize(0))));
    }

    #[test]
    fn test_decompress_invalid_packet_size_not_byte_aligned() {
        let data = vec![0u8; 10];
        let result = decompress(&data, 721, 1);
        assert!(matches!(result, Err(PocketError::InvalidPacketSize(721))));
    }

    #[test]
    fn test_decompress_invalid_robustness() {
        let data = vec![0u8; 10];
        let result = decompress(&data, 720, 8);
        assert!(matches!(result, Err(PocketError::InvalidRobustness(8))));
    }

    #[test]
    fn test_decompress_empty_input() {
        let data: Vec<u8> = vec![];
        let result = decompress(&data, 720, 1);
        assert!(matches!(result, Err(PocketError::UnexpectedEndOfInput)));
    }

    #[test]
    fn test_decompress_valid_params() {
        let data = vec![0u8; 10];
        let result = decompress(&data, 720, 1);
        assert!(result.is_ok());
    }
}
